File: tests/preview_session_test.py
#!/usr/bin/env python3
"""Exercise real dynamic-library replacement without a display or app checkout."""

import os
from pathlib import Path
import shlex
import subprocess
import sys
import tempfile


ROOT = Path(__file__).resolve().parents[1]
CC = shlex.split(os.environ.get("CC", "cc"))
HOST = r'''
#include "app_host.h"
static int value = VERSION;
static int count(void *context) { (void)context; return value; }
static void draw(void *context, Rectangle bounds) {
    (void)context; (void)bounds; value += 10;
}
static AppHost host = {.screen_count = count, .draw = draw};
AppHost *CreateAppHost(int version, const char *project) {
    (void)project;
#ifdef REJECT
    return 0;
#endif
    if(version != APP_HOST_ABI_VERSION) return 0;
#ifdef NO_DRAW
    host.draw = 0;
#endif
    return &host;
}
#ifndef NO_DESTROY
void DestroyAppHost(AppHost *value) { (void)value; }
#endif
'''
BUILD = r'''
import pathlib
import shutil
import sys
import time
import json

assert sys.argv[1] == "-C" and sys.argv[3] == "kryon-host", sys.argv
assert sys.argv[4] == "KRYON_DIR=directory with 'quotes' and $dollars", sys.argv
project = pathlib.Path(sys.argv[2])
mode = (project / "mode").read_text()
print("building " + str(project), flush=True)
if mode == "build-failure":
    print("sample.kry:4:2: invalid widget property", file=sys.stderr)
    print(json.dumps(dict(severity="error", code="check.widget", message="invalid widget property",
                          path="sample.kry", line=4, column=2, end_line=4, end_column=8)), file=sys.stderr)
    sys.exit(1)
if mode == "slow-good-2":
    time.sleep(0.25)
    mode = "good-2"
destination = project / "build/kryon/app_host.so"
destination.parent.mkdir(parents=True, exist_ok=True)
if mode == "invalid-library":
    destination.write_text("not a shared library")
else:
    shutil.copyfile(project / (mode + ".so"), destination)
'''


def run(*args, **kwargs):
    return subprocess.run(*args, check=True, **kwargs)


with tempfile.TemporaryDirectory(prefix="kryon-preview-test.") as temporary:
    work = Path(temporary)
    project = work / "project with 'quotes' and $dollars"
    project.mkdir()
    source = work / "host.c"
    source.write_text(HOST)
    for name, definitions in {
        "good-1": ["-DVERSION=1"],
        "good-2": ["-DVERSION=2"],
        "missing-symbol": ["-DVERSION=3", "-DNO_DESTROY"],
        "rejected-abi": ["-DVERSION=4", "-DREJECT"],
        "missing-draw": ["-DVERSION=5", "-DNO_DRAW"],
    }.items():
        run(CC + ["-shared", "-fPIC", "-I", str(ROOT / "include"), *definitions,
                  str(source), "-o", str(project / (name + ".so"))])
    builder = work / "make fixture"
    builder.write_text(f"#!{sys.executable}\n" + BUILD)
    builder.chmod(0o755)
    binary = work / "preview-session-test"
    run(CC + ["-std=c99", "-Wall", "-Wextra", "-Werror", "-I", str(ROOT / "include"),
              "-I", str(ROOT / "cmd/kryon-preview"),
              str(ROOT / "tests/preview_session_test.c"),
              str(ROOT / "cmd/kryon-preview/session.c"),
              str(ROOT / "cmd/kryon-preview/watch.c"),
              str(ROOT / "src/kry_std/kry_dylib.c"), str(ROOT / "src/kry_std/kry_json.c"),
              "-ldl", "-o", str(binary)])
    environment = dict(os.environ, MAKE=str(builder), TMPDIR=str(work),
                       KRYON_DIR="directory with 'quotes' and $dollars")
    result = subprocess.run([str(binary), str(project)], env=environment,
                            capture_output=True, text=True, timeout=30)
    if result.returncode:
        sys.stderr.write(result.stdout + result.stderr)
        raise SystemExit(result.returncode)
    assert "sample.kry:4:2: invalid widget property" in result.stderr
    assert not list(work.glob("kryon-preview.*")), list(work.iterdir())
    print(result.stdout.strip())
    launcher = work / "kryon"
    launcher.write_bytes((ROOT / "scripts/kryon.sh").read_bytes())
    preview = work / "kryon-preview"
    preview.write_text(f"#!{sys.executable}\nimport sys\nprint(repr(sys.argv[1:]))\n")
    preview.chmod(0o755)
    result = run(["sh", str(launcher), "--project", str(project), "preview",
                  "--source", "source with spaces.kry"], capture_output=True, text=True)
    assert result.stdout.strip() == repr(["watch", "--project", str(project),
                                         "--source", "source with spaces.kry"])
    print("preview launcher: argument forwarding without a build lock passed")

File: tests/preview_session_test.c
#define _POSIX_C_SOURCE 200809L

#include "session.h"
#include "watch.h"

#include <assert.h>
#include <string.h>
#include <time.h>
#include <sys/stat.h>
#include <unistd.h>

static void
mode(const char *project, const char *name)
{
    char filename[PREVIEW_PATH_MAX];
    FILE *file;

    snprintf(filename, sizeof(filename), "%s/mode", project);
    file = fopen(filename, "w");
    assert(file != NULL);
    assert(fputs(name, file) >= 0);
    assert(fclose(file) == 0);
}

static int
finish(PreviewSession *session)
{
    struct timespec pause = {0, 1000000};
    int result;

    for(int attempt = 0; attempt < 10000; attempt++) {
        result = PreviewPollBuild(session);
        if(result != 0)
            return result;
        nanosleep(&pause, NULL);
    }
    assert(!"preview build timed out");
    return -1;
}

static int
value(PreviewSession *session)
{
    assert(session->host != NULL);
    return session->host->screen_count(session->host->userdata);
}

int
main(int argc, char **argv)
{
    PreviewSession session = {0};
    PreviewSession other = {0};
    const char *failures[] = {"build-failure", "invalid-library", "missing-symbol",
                              "rejected-abi", "missing-draw"};
    char directory[PREVIEW_PATH_MAX];
    char library[PREVIEW_PATH_MAX];
    AppHost *original;
    uint64_t before;
    uint64_t after;
    char filename[PREVIEW_PATH_MAX];
    FILE *source;

    assert(argc == 2);
    assert(PreviewSourceStamp(argv[1], &before));
    snprintf(filename, sizeof(filename), "%s/source.kry", argv[1]);
    source = fopen(filename, "w");
    assert(source != NULL);
    fputs("Value :: 1\n", source);
    fclose(source);
    assert(PreviewSourceStamp(argv[1], &after) && after != before);
    assert(unlink(filename) == 0);
    assert(PreviewSourceStamp(argv[1], &after) && after == before);
    mode(argv[1], "good-1");
    assert(PreviewStartBuild(&session, argv[1]) > 0);
    assert(finish(&session) > 0);
    assert(value(&session) == 1);
    snprintf(filename, sizeof(filename), "%s/build/generated.kry", argv[1]);
    source = fopen(filename, "w");
    assert(source != NULL);
    fputs("generated build output", source);
    fclose(source);
    assert(PreviewSourceStamp(argv[1], &after) && after == before);
    session.host->draw(session.host->userdata, (Rectangle){0});
    assert(value(&session) == 11);
    original = session.host;
    strcpy(library, session.library_path);
    for(size_t i = 0; i < sizeof(failures) / sizeof(failures[0]); i++) {
        mode(argv[1], failures[i]);
        assert(PreviewStartBuild(&session, argv[1]) > 0);
        assert(session.host == original);
        assert(finish(&session) < 0);
        if(i == 0) {
            assert(session.diagnostic_count == 1);
            assert(strcmp(session.diagnostics[0].path, "sample.kry") == 0);
            assert(strcmp(session.diagnostics[0].code, "check.widget") == 0);
            assert(session.diagnostics[0].line == 4 && session.diagnostics[0].column == 2);
            assert(session.diagnostics[0].end_column == 8);
        } else {
            assert(session.diagnostic_count == 0);
        }
        assert(session.host == original);
        assert(value(&session) == 11);
        assert(access(library, F_OK) == 0);
    }
    mode(argv[1], "slow-good-2");
    assert(PreviewStartBuild(&session, argv[1]) > 0);
    assert(PreviewPollBuild(&session) == 0);
    session.host->draw(session.host->userdata, (Rectangle){0});
    assert(value(&session) == 21);
    assert(finish(&session) > 0);
    assert(value(&session) == 2);
    assert(access(library, F_OK) != 0);
    assert(session.error[0] == '\0');

    mode(argv[1], "good-1");
    assert(PreviewStartBuild(&other, argv[1]) > 0);
    assert(finish(&other) > 0);
    assert(strcmp(session.directory, other.directory) != 0);
    assert(strcmp(session.library_path, other.library_path) != 0);
    PreviewClose(&other);
    assert(value(&session) == 2);

    strcpy(directory, session.directory);
    mode(argv[1], "slow-good-2");
    assert(PreviewStartBuild(&session, argv[1]) > 0);
    PreviewClose(&session);
    assert(session.build_pid == 0 && session.host == NULL);
    assert(access(directory, F_OK) != 0);
    PreviewClose(&session);
    puts("preview session: failure preservation, responsive build, recovery, isolation, cleanup passed");
    return 0;
}
